kubetest/eks/eks.go

```diff
@@ -0,0 +1,237 @@
+type tester struct {
+	stopc         chan struct{}
+	cfg           *eksconfig.Config
+	awsTesterPath string
+	kubectlPath   string
+}
+
+// New creates a new EKS deployer with AWS CLI.
+func New(cfg *eksconfig.Config) (eksdeployer.Interface, error) {
+	cfg.Embedded = false
+	tr := &tester{
+		stopc: make(chan struct{}),
+		cfg:   cfg,
+	}
+	var err error
+	tr.awsTesterPath, err = kexec.New().LookPath("awstester")
+	if err != nil {
+		return nil, fmt.Errorf("cannot find 'awstester' executable (%v)", err)
+	}
+	tr.kubectlPath, err = kexec.New().LookPath("kubectl")
+	if err != nil {
+		return nil, fmt.Errorf("cannot find 'kubectl' executable (%v)", err)
+	}
+	_, err = kexec.New().LookPath("aws-iam-authenticator")
+	if err != nil {
+		return nil, fmt.Errorf("cannot find 'aws-iam-authenticator' executable (%v)", err)
+	}
+	return tr, nil
+}
+
+var (
+	timeout = 3 * time.Hour
+	verbose = true
+	control = process.NewControl(timeout, time.NewTimer(timeout), time.NewTimer(timeout), verbose)
+)
+
+// Up creates a new EKS cluster.
+func (tr *tester) Up() (err error) {
+	createCmd := exec.Command(
+		tr.awsTesterPath,
+		"eks",
+		"--path="+tr.cfg.ConfigPath,
+		"create",
+		"cluster",
+	)
+	errc := make(chan error)
+	go func() {
+		_, oerr := control.Output(createCmd)
+		errc <- oerr
+	}()
+	select {
+	case <-tr.stopc:
+		fmt.Fprintln(os.Stderr, "received stop signal, interrupting 'create cluster' command...")
+		ierr := createCmd.Process.Signal(syscall.SIGINT)
+		err = fmt.Errorf("'create cluster' command interrupted (interrupt error %v)", ierr)
+	case err = <-errc:
+	}
+	return err
+}
+
+// Down tears down the existing EKS cluster.
+func (tr *tester) Down() (err error) {
+	_, err = control.Output(exec.Command(
+		tr.awsTesterPath,
+		"eks",
+		"--path="+tr.cfg.ConfigPath,
+		"delete",
+		"cluster",
+	))
+	return err
+}
+
+// IsUp returns an error if the cluster is not up and running.
+func (tr *tester) IsUp() (err error) {
+	_, err = control.Output(exec.Command(
+		tr.awsTesterPath,
+		"eks",
+		"--path="+tr.cfg.ConfigPath,
+		"check",
+		"cluster",
+	))
+	if err != nil {
+		return err
+	}
+
+	tr.cfg, err = eksconfig.Load(tr.cfg.ConfigPath)
+	if err != nil {
+		return err
+	}
+
+	if tr.cfg.ClusterState.Status != "ACTIVE" {
+		return fmt.Errorf("cluster %q status is %q",
+			tr.cfg.ClusterName,
+			tr.cfg.ClusterState.Status,
+		)
+	}
+	return nil
+}
+
+// TestSetup checks if EKS testing cluster has been set up or not.
+func (tr *tester) TestSetup() error {
+	return tr.IsUp()
+}
+
+// GetClusterCreated returns EKS cluster creation time and error (if any).
+func (tr *tester) GetClusterCreated(v string) (time.Time, error) {
+	err := tr.IsUp()
+	if err != nil {
+		return time.Time{}, err
+	}
+	tr.cfg, err = eksconfig.Load(tr.cfg.ConfigPath)
+	if err != nil {
+		return time.Time{}, err
+	}
+	return tr.cfg.ClusterState.Created, nil
+}
+
+// DumpClusterLogs uploads local cluster logs to S3.
```

Comment:
humm our infra already supports upload logs to gcs, which works with gubernator & testgrid etc.. I would dump the log to local $ARTIFACTS and let our tooling layer handle that, otherwise you can also choose to support make test-infra handle artifacts from s3.

Reply:
We want to use the GCS first. And then migrate later, if needed. Will change the code to dump logs to $ARTIFACTS, and address other comments within this week.

> choose to support make test-infra handle artifacts from s3

Sounds good. We can write some proposal later to do this.

Thanks!